daemon/algod/api/server/v2/handlers.go

```diff
@@ -67,6 +68,144 @@ type NodeInterface interface {
+			VoteFirstValid:  uint64(record.FirstValid),
+			VoteLastValid:   uint64(record.LastValid),
+			VoteKeyDilution: record.KeyDilution,
+		},
+	}
+
+	// These are pointers but should always be present.
+	if record.Voting != nil {
+		participationKey.Key.VoteParticipationKey = record.Voting.OneTimeSignatureVerifier[:]
+	}
+	if record.VRF != nil {
+		participationKey.Key.SelectionParticipationKey = record.VRF.PK[:]
+	}
+
+	// Optional fields.
+	participationKey.EffectiveFirstValid = roundToPtrOrNil(record.EffectiveFirst)
+	participationKey.EffectiveLastValid = roundToPtrOrNil(record.EffectiveLast)
+	participationKey.LastVote = roundToPtrOrNil(record.LastVote)
+	participationKey.LastBlockProposal = roundToPtrOrNil(record.LastBlockProposal)
+	participationKey.LastVote = roundToPtrOrNil(record.LastVote)
+	participationKey.LastStateProof = roundToPtrOrNil(record.LastStateProof)
+
+	// Special case for first valid on round 0
+	if record.EffectiveLast != 0 && record.EffectiveFirst == 0 {
+		zero := uint64(0)
+		participationKey.EffectiveFirstValid = &zero
+	}
+
+	return participationKey
+}
+
+// GetParticipationKeys Return a list of participation keys
+// (GET /v2/participation)
+func (v2 *Handlers) GetParticipationKeys(ctx echo.Context) error {
+	partKeys, err := v2.Node.ListParticipationKeys()
+
+	if err != nil {
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
+
+	var response []generated.ParticipationKey
+
+	for _, participationRecord := range partKeys {
+		response = append(response, convertParticipationRecord(participationRecord))
+	}
+
+	return ctx.JSON(http.StatusOK, response)
+}
+
+// AddParticipationKey Add a participation key to the node
+// (POST /v2/participation)
+func (v2 *Handlers) AddParticipationKey(ctx echo.Context) error {
+
+	buf := new(bytes.Buffer)
+	_, err := buf.ReadFrom(ctx.Request().Body)
+	if err != nil {
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
+	partKeyBinary := buf.Bytes()
+
+	if len(partKeyBinary) == 0 {
+		err := fmt.Errorf(errRESTPayloadZeroLength)
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
+
+	partID, err := v2.Node.InstallParticipationKey(&partKeyBinary)
+
+	if err != nil {
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
+
+	response := generated.PostParticipationResponse{PartId: partID.String()}
+	return ctx.JSON(http.StatusOK, response)
+
+}
+
+// DeleteParticipationKeyByID Delete a given participation key by id
+// (DELETE /v2/participation/{participation-id})
+func (v2 *Handlers) DeleteParticipationKeyByID(ctx echo.Context, participationID string) error {
+
+	decodedParticipationID, err := account.ParticipationIDFromString(participationID)
+
+	if err != nil {
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
+
+	err = v2.Node.RemoveParticipationKey(decodedParticipationID)
+
+	if err != nil {
+
+		if errors.Is(err, account.ErrParticipationIDNotFound) {
+			return ctx.JSON(http.StatusOK, generated.ErrorResponse{Message: "participation id not found"})
```

Comment:
This doesn't sounds like a "StatusOK" issue to me.. shouldn't it be `http.StatusNotFound` ?

Reply:
Stephen and I had the same discussion and this is what we decided on. I don't think there's a good single answer.

On one hand: I requested a delete state change. If there is nothing to delete there was no state change so return an error.

On the other: I want to make sure that the specified object does not exist in the system. The fact it doesn't exist at all is irrelevant, the final state is still what was requested.

Reply:
I think the answer is in the REST definitions; look in https://www.restapitutorial.com/lessons/httpmethods.html : 
for the DELETE command, the response should be : 
200 (OK). 
404 (Not Found), if ID not found or invalid.

Reply:
Interesting, this page goes into more detail further down:
```
There is a caveat about DELETE idempotence, however. Calling DELETE on a resource a second time will often return a 404 (NOT FOUND) since it was already removed and therefore is no longer findable. This, by some opinions, makes DELETE operations no longer idempotent, however, the end-state of the resource is the same. Returning a 404 is acceptable and communicates accurately the status of the call.
```

I see a lot of debate on the subject but the majority seem to side with the 404 approach, so I'll change it.